### ISSUE TITLE
Think how to move files for new archi and deprecate the rest

### ISSUE THREAD

**Issue by hugtalbot** [sofa-framework/sofa#346]

**Comment by damienmarchal** [comment]
Hi Hugo, This is a very challenging target.

I would probably try something something like:

1-  For release 17.12
- move all the "not core" files in plugins. 
- for each "SofaXX/blahblah.h" in the old location have this kind of file with:
```cpp
/// This file is now deprecated and have been moved in the "SofaXX" plugin. You need to
/// update your code to #include <SofaConstraint/blabla.h> 
#ifdef SMOOTH_TRANSITION_UNTIL_17_12
   #include <applications/plugins/SofaXX/blahblah.h>
#endif
```
- for each component you really want to deprecate add a hook in the SceneChecker saying "component xxYYZZ is now deprecated and will be removed in sofa 18.06. Use alternative XX and if there is none please contact us open Issue in our bugtracker with tag 'sofaNG'. "

Then in 18.06...really removes the component and update the SceneChecker hook so that it simply print somehing like "this component have been removed since sofa 18.06. Alternative component is now XXXX." 

Of course the more automated it will be through script the best it will be.

**Comment by damienmarchal** [comment]
Not a lot of contribution here. 

In recent PR I'm doing moving code from modules to plugins. 
I think that the followin approach combining plugins & packages (group of selected plugin or components)
plugins:
    - plugin1
    - plugin2
    - plugin3
    - plugin4

packages:
    - SofaAdvanced (include pluin1, plugin2)
    - SofaMisc (include plugin3, plugin4)
    - SofaNG(include plugin1, and MechanicalFEM)
....

What do you think ?

**Comment by damienmarchal** [comment]
Hi all, 

At Defrost we are now moving most of the module into plugins to accelerate sofa build.

**Comment by damienmarchal** [comment]
This include PR #389 #407  #396 
It would be nice to have feedback..otherwise...well after merge it will be to late ;)

**Comment by damienmarchal** [comment]
Here is another contribution to handle in an organized and smooth way sofa's components lifetime and deprecation.

The idea is to have three sets of components in the CMakeList.txt. 
- the first sets contains more or less the SofaNG components.
- the second sets contains all existing sofa components. 
- the third sets contains deprecated components. 

Selecting which set to compile is controlled using a CMakeList argument in an inclusive way (AllComponent include SofaNG while deprecated includes allComponent) 

When a component is in the deprecated set we should add the corresponding line to the 
SceneChecker to warn users in a friendly way.

This approach is complementary to the pluginization of module which allow user to control  the kinds of feature they need and don't compile the others. 

What do you think about it ?

**Comment by untereiner** [comment]
For me handling different sets of sofa components with a lots of CMakeLists is cumbersome. 
I am not ashamed to say that I do not understand the difference between SofaKernel/SofaFramework/SofaMisc/SofaNG/SofaAdvanced/SofaGeneral

I would suggest an another direction more in the spirit of Qt, e.g. having meaningful modules. 
In Qt they have: Qt Core/ Qt Multimedia / Qt Network etc. 
I suggest to have something similar: Sofa Core (with only the glue) / Sofa Solvers / Sofa Elastic / Sofa Fluid / Sofa Collision /
In each of these modules we could have several algorithms from the simplest and cheap to the more complex expensive one.

**Comment by damienmarchal** [comment]
Thank for your feedback, 

As you I don't understand the SofaKernel/SofaFramework stuff and I'm always searching for the right component in the wrong place so I fully support you on that, the reason is that Misc vs Advanced vs NG have different meaning for different person so it is a totally arbitrary segmentation. 

But I'm not really talking about this here. What I have in mind is more or less orthogonal to this aspect (and it is fully compatible with the existing scheme or one you could define in the future grouping stuff like Qt). What I would like to discuss is more about how do we management the components life-time (from integration to deprecation) and not where to put them. I'm not sure I'm very clear sorry for that. 

Here is our working branch: 
https://github.com/SofaDefrost/sofa/tree/deprecating_components

You can see there that there is no more CMakeLists than what we have now (or what we could have in some future if some serious cleaning and refactoring is done to organize the component 'à la Qt'...but I have no time for that ;)). 

The goods I see of the proposed approach, in addition to be orthogonal to the other concerns, is that it is very progressive. It change nothing for existing user of Sofa while at the same time we can progressively deprecated components from Sofa by just moving them from one set to one other (and updating the scenechecker). In a symetric way we can also progressively integrate in the core set of components the one that have been "elected" by the SofaNG workgroup. And for the other component...well nothing change. 

My dream is that by doing so we could small step after small step, remove 50% of the components that should be deprecated and that, combined with the pluginization lighten sofa.

**Comment by StephaneCotin** [comment]
Hi Damien

I understand what you want to do (I think) and it's certainly an important aspect of cleaning SOFA. If by moving a component to a specific location we change its 'status' to 'deprecated' (and the user receives the appropriate warning), it seems a very good solution.
However, I did not understand how we'll do this in practice with SOFA-NG. In the many components that would go in plugins, some of them need to be flagged as 'deprecated' and others as 'too buggy for SOFA-NG', i.e. how / when do we decide where a component sits on its own life cycle? Knowing that I would be very tempted to put many of these in a 'close to retirement' state ;-)

**Comment by untereiner** [comment]
my fault, I didn't understand your point at the first time. I agree with you that managing the deprecation with specific sets in a CMakeLists can be efficient.
In my mind the place you put the code and its management are somehow connected especially when lighten "sofa" because nobody is defining what sofa is/contains/is reduced to. 
More I think about more I think that there is no sofa core or whatever. There are files which have a meaning to be together.
Instead of removing files/classes/components I suggested to think about moving files/classes/components and then evaluate their stage of deprecation.

**Comment by damienmarchal** [comment]
Hi Stéphane, thanks for the feedback, 

If we split the existing set of components into sub-sets eg ("stable", "fullset", "deprecated") our starting point would be the current sofa with everything in the "fullset". Starting from  that we can start moving the good enough one in the "minimal" or "stable" set, the bad ones into the "deprecated" set  and keep the one we cannot decide into the "fullset". 

You are totally right that this proposal is not providing any ideas on who will take the decision to move one component into the deprecated set. This is very interesting question but it deserve a whole issue & STC or an autocratic gentle decision :)

To me the presented approach is interesting as:
- it is incremental and thus we don't have to wait any decision of a definitive list of components that have to be keep. We can already, small step after small step, moving the ones for which there is a agreement into the right set (and people not using the set will stop being bothered with them). 
- it is bi-directionnal, so people can work in parallel either to move the components into "stable" or into "deprecated".  So team looking for the "future" should put their time on making a good "stable set"
- the approach also cares about users because until a component is removed from the deprecated set" a Sofa version can still be compiled with the deprecated components. 
- it is reversible, and each time a component is put into the "deprecated" set users of the component are warned when loading their scene, and thus, they can complain to sofa-dev that we shouldn't have deprecated a specific component (or that they are going to invest themself to maintain it). 
- it is complementary to organizing component sharing functionalities into module or plugins or packages (group of plugins) which also help to reduce the compilation time of sofa. 

@untereiner I think in the existing sofa we have used component grouping (with modules or plugins) to handle lifetime management & functionally grouping & level of quality, which are different concerns, with a single tool (the modules). 
As you are proposing I dream of a clear re-organization of the different modules/plugins but I this sound very ambitious to me and I remember of a meeting 2 years ago where this was already discussed...and nothing moved. My assumption is that this is because the step is so big that no-one can afford it. 

This is why I'm trying to find less ambitious but more incremental approach. 

Eg with the current approach if you have 1 hour time to offer you can: 
- locate the leaves of the #include graph (these are component that have low coupling and thus are easy to remove)
- move the file to the deprecated set (if you think it should)
- add the component into the scene-checker to warn users.  
- make a PR
- if the PR is accepted you can be proud of the result because you saved:
```
       1 second (average compilation time per component)
*    20  (number of sofa compilation per day)
*  150  (number of sofa developpers)
= 3000 seconds per day. 
```
The return of investment is very high :)

**Comment by damienmarchal** [comment]
There is also a big challenge connected to deprecating/SofaNG and moving component. It is cutting the edge in the dependency graph. Some edge are of course legitimate...but for plenty of them this is the result of either:
- putting method definition in headers (which thus require the include of used components)
- required for template instantiation

This make every code change a real nightmare.So buy a chainsaw and submit PR to cut into this forest of dependencies (again it is better to locate the object that are easily fix).

**Comment by fredroy** [comment]
Closing as SofaNG and its compat layer have been implemented.
